xdg-activation: deny activation if the token is incomplete

alebastr opened the pull request:
Where incomplete means that the requesting app did not provide a focused surface or a valid input event serial.
The token is still valid, so set the urgency instead.

See also #7446

***
This works exactly as I imagined. With `for_window [app_id=xxx] focus_on_window_activation focus`, the following happens:
A self-issued token without a serial (`QWindow::alert()` in Qt 6.6) can set urgency but does not switch workspace or bring the window forward.
A token from a click on mako notification allows the window to activate itself.

I'm not happy with the abuse of `launcher_ctx` (which may be not set) for an unrelated flag. Really don't see a way to avoid this, because it's too late to check if the token was fully validated in `request_activate`.

alebastr commented:
A bit of a missing context here is that xdg-activation is really overloaded. It's being used as a
* startup notification protocol for new surfaces
* urgency request protocol (as `QWindow::alert()` and foot's `wayl_win_set_urgent` are doing)
* actual activation of another surface

It's not specified by the protocol, but other implementations settled on the following semantics:

[Mutter](https://gitlab.gnome.org/GNOME/mutter/-/blob/e928128743f0ada91543cb548a56fd4676269dc6/src/wayland/meta-wayland-activation.c#L289) checks in the `request_activate` handler that the token->seat and the token->surface are available and the serial is good. If the check passes, it's treated as an activation request, otherwise it's an urgency request.

[KWin](https://invent.kde.org/plasma/kwin/-/blob/cc2e6fc96cbed230b4cdcd2353a805ec26dd2553/src/xdgactivationv1.cpp#L117) goes one step above and check that the surface that issued the token is active.

***

For sake of this pull request, it would be good to define
* what should distinguish activation from urgency
* how that should interact with `focus_on_window_activation`
* should any of these checks affect startup notifications

Where the first point is sort of security related: what restrictions should we put on a window before allowing it to bring focus to itself or anything else. Current behavior is no restrictions with `focus_on_window_activation focus`/same workspace with `focus_on_window_activation smart`.

After checking the other compositors, I'm going to propose the following model:
1. Startup notifications should continue working as is.
2. Activation request handler should determine if the token can activate a window, by checking that `token->seat` and `token->surface` are present. It would mean that _at the moment of token issue_, the surface had focus and used a valid serial as checked by wlroots, and that both the seat and the surface that requested the token are still available. The result of this step is a decision whether we want to proceed with activation or request urgency.
   (note that it is stricter than the current diff, but more relaxed than KWin/Mutter)
4. `focus_on_window_activation` value will be consulted to determine if activation or urgency request should be downgraded or denied. Just as it is already implemented.

alebastr commented:
>     2. Activation request handler should determine if the token can activate a window, by checking that `token->seat` and `token->surface` are present. It would mean that _at the moment of token issue_, the surface had focus and used a valid serial as checked by wlroots, and that both the seat and the surface that requested the token are still available. The result of this step is a decision whether we want to proceed with activation or request urgency.
>        (note that it is stricter than the current diff, but more relaxed than KWin/Mutter)

Hm. That may be still too strict and could introduce races. E.g. in the case of a desktop notification, notification daemon will pass the token over DBus and destroy the surface. Now, if the app does not manage to send `request_activate` in time it won't get activated.
And once you remove `and that both the seat and the surface that requested the token are still available` from the wording, we'll get the behavior from the first commit.

rpigott commented on sway/xdg_activation_v1.c:
```diff
@@ -44,7 +49,13 @@ void xdg_activation_v1_handle_new_token(struct wl_listener *listener, void *data
 	if (ws) {
-		launcher_ctx_create(token, &ws->node);
+		struct launcher_ctx *ctx = launcher_ctx_create(token, &ws->node);
+		if (ctx) {
```
As above, but first move this into launcher_ctx create. There you can record whatever necessary state.

alebastr replied:
The assignment was here because it's an xdg-activation policy decision and hiding it in the `launcher_ctx_create` didn't seem right.
Well, if we go with separate flags for seat/surface, it would be reasonable to set the flags in `launcher_ctx_create`, but do the check in `xdg_activation_v1_handle_request_activate`.

rpigott replied:
> The assignment was here because it's an xdg-activation policy decision

`can_activate` is a policy decision, but something like `had_focused_surface` is a plain fact about the compositor state. That is what I'm suggesting.

alebastr replied:
Ok, latest update does that.

rpigott reviewed (changes requested):
Proposal sounds reasonable to me. Personally I think I'd like it if manually changing focus before activation cancelled automatic focus switches, but we don't have to implement that here.

alebastr commented on include/sway/desktop/launcher.h:
```diff
@@ -11,6 +11,8 @@ struct launcher_ctx {
 
 	bool activated;
+	// The token can be used to activate a window
+	bool can_activate;
```
> just record the information you need to derive it at the time of activation

The pointers to a seat/surface provided when requesting the token can become stale. Seems like too much of a footgun to store these in the launcher context.

Replacing these with flags (`TOKEN_HAS_FOCUSED_SURFACE`/`TOKEN_HAS_VALID_SERIAL`) could be an alternative. But then it sounds like something to be done in the token struct on wlroots side, like Simon proposed on IRC a while ago. And I'd like to see some variant of this change in v1.9.

alebastr commented:
Reverted to the initial version and rewrote the commit message.

> Personally I think I'd like it if manually changing focus before activation cancelled automatic focus switches, but we don't have to implement that here.

I'd agree with that, but the case of freedesktop notifications bothers me. It would be impossible to raise a window by clicking the notification, if we check that the notification surface exists and has focus. At least mako dismisses the notification immediately after sending the DBus signals.

alebastr commented on sway/xdg_activation_v1.c:
```diff
@@ -32,9 +32,14 @@ void xdg_activation_v1_handle_request_activate(struct wl_listener *listener,
-	struct wlr_seat *wlr_seat = event->token->seat;
-	struct sway_seat *seat = wlr_seat ? wlr_seat->data : NULL;
-	view_request_activate(view, seat);
+	if (ctx == NULL || ctx->can_activate) {
```
So this should be `if (ctx != NULL && ctx->can_activate)`? Got it, will fix with the next update.

rpigott commented:
> It would be impossible to raise a window by clicking the notification, if we check that the notification surface exists and has focus.

Right, but instead of that we could in principal refuse to honor any focus activation whose token was generated before the last time the user manually changed focus. That would have to be handled in the ipc commands/seatops that manually change focus. This way shifting focus down the focus stack after the notification window disappears wouldn't interrupt any pending focus hand-off, but manually switching windows would.

Again, we don't have to implement that here. I'm just speculating on what kind of compositor policy might be sensible.

alebastr commented on sway/xdg_activation_v1.c:
```diff
@@ -32,9 +32,14 @@ void xdg_activation_v1_handle_request_activate(struct wl_listener *listener,
-	view_request_activate(view, seat);
+	if (ctx != NULL && ctx->had_focused_surface && ctx->had_valid_serial) {
+		struct wlr_seat *wlr_seat = event->token->seat;
+		struct sway_seat *seat = wlr_seat ? wlr_seat->data : NULL;
```
Makes sense to me, done. Also removed the now unused `had_valid_serial` flag.

rpigott commented:
N.B. rebased on #7962.

LGTM, thanks.

rpigott commented:
It should if Firefox requests focus. That means the client that opens a link needs to send it the token though. I'm guessing Firefox doesn't implement it anyway...

 @alebastr do you know what other compositors do here? Does this focus change occur? I wonder if they hack the focus change into whatever xdg-open calls.

alebastr commented:
The focus works if a client with focus and activation support (e.g.
Thunderbird) generates an activation token and sends it to Firefox. I've
just verified that it still behaves as expected.

What doesn't work is xdg-open https://url, because there's no requesting
surface, and the request gets downgraded to urgency. And that's how it
works in GNOME, KDE, etc.
If we relax that check, we'll get to the initial state, where an
application expects to set urgency (foot on BEL) but gets instant focus.
FOWA is a global flag and cannot be used per window to mitigate that :(

With GNOME there's a notification 'Firefox has opened the url, now click me
to focus the window', but I have no idea which component sends it. Most
likely, Firefox itself.

On Tue, Feb 27, 2024 at 10:33 AM Ronan Pigott ***@***.***>
wrote:

> It should if Firefox requests focus. That means the client that opens a
> link needs to send it the token though. I'm guessing Firefox doesn't
> implement it anyway...
>
> @alebastr <https://github.com/alebastr> do you know what other
> compositors do here? Does this focus change occur? I wonder if they hack
> the focus change into whatever xdg-open calls.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/swaywm/sway/pull/7880#issuecomment-1967361310>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAUOSP2TLIZI2Y42IUM2XVDYVYRJJAVCNFSM6AAAAABBEACLLWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNRXGM3DCMZRGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
With best regards,
Aleksei Bavshin

rpigott commented:
Good to know Firefox can receive tokens somehow, but what out of band method does it implement to receive them? I thought Firefox opened new tabs with dbus, but just inspecting the interface it doesn't look like it can receive them there — it takes one argument which I assume is the uri.
